torch/distributed/checkpoint/state_dict.py

```diff
@@ -332,6 +344,130 @@ def _init_optim_state(optim: torch.optim.Optimizer) -> None:
     optim.zero_grad(set_to_none=True)
 
 
+def _flatten_optim_state_dict(value: Any, prefix: str = "") -> Dict[str, ValueType]:
+    ret = {}
+    if isinstance(value, dict):
+        for key, child_value in value.items():
+            next_prefix = f"{prefix}.{key}" if prefix else key
+            ret.update(_flatten_optim_state_dict(child_value, next_prefix))
+    elif isinstance(value, (DTensor, ShardedTensor, torch.Tensor, int, float)):
+        ret[prefix] = value
+    elif isinstance(value, (list, tuple)):
+        assert len(value) > 0, prefix
```

Comment:
Why aren't we okay with empty lists/tuples?

Also, tangentially, I wonder if pytrees will help simplify here haha